kafka/channel/cmd/channel_dispatcher/main.go

```diff
@@ -35,5 +35,8 @@ func main() {
 		ctx = injection.WithNamespaceScope(ctx, ns)
 	}
 
-	sharedmain.MainWithContext(ctx, component, controller.NewController)
+	cfg := sharedmain.ParseAndGetConfigOrDie()
+
+	ctx = sharedmain.WithHADisabled(ctx)
```

Comment:
is HA disabled required ?

Reply:
this is not required.

Reply:
According to my understanding, HA is enabled by default, see https://github.com/knative/eventing-contrib/blob/ba193081c89e71f1f9a4bbc10c774df65d079e4b/kafka/channel/pkg/client/injection/reconciler/messaging/v1beta1/kafkachannel/controller.go#L52 and https://github.com/knative/pkg/blob/c4576fd38ec2d788ab3a6ae55527d7fffb865c5d/injection/sharedmain/main.go#L145, and on the dispatcher side, having HA enabled means with don't get `ReconcileKind` calls for each replica, see https://github.com/knative/eventing-contrib/blob/ba193081c89e71f1f9a4bbc10c774df65d079e4b/kafka/channel/pkg/client/injection/reconciler/messaging/v1beta1/kafkachannel/reconciler.go#L193

Reply:
@vaikas @mattmoor, Could you please take a look at this?

I'm thinking that a better approach is to use both `ObserveKind` and `ReconcileKind`.
`ObserveKind` instantiates consumers and `ReconcileKind` instantiates consumers and sets the status of the channel.

WDYT?

Reply:
Just to make sure I understand. For the kafka channel controller, you need to use ReconcileKind, because you are actually doing work and will need to do things and set status, etc.? Correct?

For the dispatcher, you just need to be aware of all the Channels so that you can see the channels and fire up dispatch functions, but you do not need to do any actual reconcile work? Furthermore, you also would like to scale the dispatchers up and you want to use ObserveKind so that each dispatcher replica will be aware of the Channels it is "leading" and if you use ReconcileKind there's some extra unnecessary work, in HA mode the pkg/reconciler will not call your methods unless it is the leader?

So, yeah, that sounds right to me:
Controller - ReconcileKind
Dispatcher - ObserveKind

@n3wscott

Reply:
> Just to make sure I understand. For the kafka channel controller, you need to use ReconcileKind, because you are actually doing work and will need to do things and set status, etc.? Correct?

Correct

> For the dispatcher, you just need to be aware of all the Channels so that you can see the channels and fire up dispatch functions, but you do not need to do any actual reconcile work? Furthermore, you also would like to scale the dispatchers up and you want to use ObserveKind so that each dispatcher replica will be aware of the Channels it is "leading" and if you use ReconcileKind there's some extra unnecessary work, in HA mode the pkg/reconciler will not call your methods unless it is the leader?

Yeah, the problem is that the dispatcher reconciler changes the status:
 https://github.com/knative/eventing-contrib/blob/312de090fb65b61adf9544a93cc29e7198c82c42/kafka/channel/pkg/reconciler/dispatcher/kafkachannel.go#L173
so I thought to use a mixed approach outline above:

> I'm thinking that a better approach is to use both ObserveKind and ReconcileKind.
ObserveKind instantiates consumers and ReconcileKind instantiates consumers and sets the status of the channel.

To clarify, the leader changes the `SubscribableStatus` and fires up dispatch functions, and non-leaders only fire up dispatch functions.

Or

As Ville wrote, the dispatcher doesn't change the channel status at all, it only fires up dispatch functions, and we need to move the `SubscribableStatus` update to the controller. (This is what I do in `eventing-kafka-broker`)

Reply:
Every reconciler should have a ReconcileKind method, even if it is readonly.

ObserveKind exists to give non-leader controllers the ability to build up their own indices (think "read replicas"), but this will NOT be called on leaders, so (skimming) this doesn't sound right.

Want to discuss in HackyHour in 35?